### PR TITLE
Update Swift to use SwiftPM build system

### PR DESCRIPTION
**Pull request by externl**
This PR updates the Swift build system to use SwiftPM (https://www.swift.org/documentation/package-manager/).

## Package.swift
The primary Ice for Swift build file is the root-level `Package.swift` file. SwiftPM requires a root level file. This provides the Ice for Swift libraries (Ice, Glacier2, etc.) as well as a new SwiftPM plugin, `CompileSlice`, that will generate Swift sources from Slice. 

`CompileSlice` requires users to create `slice-plugin.json` files (any better name suggestions?) describing the Slice compilation they wish to perform. The `protobuf` plugin does something similar. `.ice` files can't be added as sources, but JSON files can. This is why we have the json file and how the plugin figures out what to compile.

Example:
```
{
  "sources": ["../../../slice/IceGrid"],
  "search_paths": ["../../../slice"]
}
```

The build relies on the generated `slice2cpp` sources since SwiftPM does not currently support `.h|.c|.cpp` source generation. So as of right now you'll need a src C++ build. Swift also performs a C++ build itself using the pre-generated sources.  In a follow-up PR our C++ build should generate `xcframework`s that we can use instead. The final version we tag and ship will need to link with an Ice for C++ `xcframework` from Homebrew.

## Testing
The tests are also built using SwiftPM, but not as part of the main build. Instead we use a `Package.swift` located in `swift/test.

Compared to the previous setup, we no longer have one bundle per test, but one library containing all the tests. 

```
❯ swift run TestDriver Slice_escape Client
Building for debugging...
[1/1] Write swift-version--58304C5D6DBC2206.txt
Build complete! (0.16s)
testing operation name... ok
testing types... ok
```

The iOS testing uses an actual app and for that we still require an Xcode project. I'm not generating it as it's only a few files and should be pretty static. No need to keep the ruby dependency just for this.

Build and running the Swift tests in iOS also currently disabled. We were not running it previously and there's a build issue I still need to fix in a followup PR.

### CODE REVIEW

File: Package.resolved [hunk 1]
```diff
@@ -0,0 +1,23 @@
+{
+  "pins" : [
+    {
+      "identity" : "mcpp",
+      "kind" : "remoteSourceControl",
+      "location" : "https://github.com/zeroc-ice/mcpp.git",
+      "state" : {
+        "branch" : "master",
+        "revision" : "bfd4b159d9428c983e2abb48c8d344003a019caf"
```
Comment (bernardnormier) [inline Package.resolved:9]: We should use a label.

File: swift/src/IceImpl/Communicator.mm [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
-#import "IceUtil.h"
-#import "ImplicitContext.h"
-#import "Logger.h"
-#import "ObjectAdapter.h"
-#import "ObjectPrx.h"
-#import "Process.h"
-#import "Properties.h"
-#import "PropertiesAdmin.h"
-#import "UnsupportedAdminFacet.h"
+#import "include/Communicator.h"
```
Comment (bernardnormier) [inline swift/src/IceImpl/Communicator.mm:2]: What goes into include vs what stays?

It's not immediately clear to me. And should this be documented in a README?

File: swift/src/CompileSlicePlugin/CompileSlicePlugin.swift [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+enum PluginError: Error {
+    case invalidTarget(Target)
+    case missingCompiler(String)
+    case missingConfigFile(String, String)
+    case missingIceSliceFiles(String)
+
+    var description: String {
+        switch self {
+        case .invalidTarget(let target):
+            return "Expected a SwiftSourceModuleTarget but got '\(type(of: target))'."
```
Comment (bernardnormier) [inline swift/src/CompileSlicePlugin/CompileSlicePlugin.swift:15]: I believe you can simplify and remove "return"

File: swift/src/CompileSlicePlugin/CompileSlicePlugin.swift [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        case .missingConfigFile(let path, let target):
+            return
+                "Missing config file '\(path)` for target `\(target)`. '. This file must be included in your sources."
+        case .missingIceSliceFiles(let path):
+            return "The Ice slice files are missing. Expected location: `\(path)`"
+        }
+    }
+}
+
+/// Represents the contents of a `slice-plugin.json` file
```
Comment (bernardnormier) [inline swift/src/CompileSlicePlugin/CompileSlicePlugin.swift:27]: We need a better documentation for this CompileSlicePlugin here, and also in a README file.

**Review by bernardnormier — APPROVED** [review]

File: .github/workflows/docs.yml [hunk 5]
```diff
@@ -48,38 +48,38 @@ jobs:
...
-          $(xcrun --find docc) process-archive \
-            transform-for-static-hosting /tmp/docbuild/Build/Products/Debug/IceGrid.doccarchive \
-            --hosting-base-path /api/ice/main/swift/icegrid \
-            --output-path docs/icegrid;
-
-          $(xcrun --find docc) process-archive \
-            transform-for-static-hosting /tmp/docbuild/Build/Products/Debug/IceStorm.doccarchive \
-            --hosting-base-path /api/ice/main/swift/icestorm \
-            --output-path docs/icestorm;
+      # - name: Generate Documentation for Swift
```
Comment (pepone) [inline .github/workflows/docs.yml:51]: add a TODO to enable Swift docs?

File: swift/src/CompileSlicePlugin/CompileSlicePlugin.swift (hunk 4 above)
Comment (externl) [reply swift/src/CompileSlicePlugin/CompileSlicePlugin.swift:27]: Yes.

File: swift/src/CompileSlicePlugin/CompileSlicePlugin.swift (hunk 3 above)
Comment (externl) [reply swift/src/CompileSlicePlugin/CompileSlicePlugin.swift:15]: Compiler doesn't like it.

File: Package.resolved (hunk 1 above)
Comment (externl) [reply Package.resolved:9]: We can once we tag a new `mcpp` release.

File: Package.swift [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        ),
+        .plugin(
+            name: "CompileSlice",
+            capability: .buildTool(),
+            dependencies: ["slice2swift"],
+            path: "swift/src/CompileSlicePlugin"
+        ),
+    ],
+    swiftLanguageVersions: [SwiftVersion.v5],
+    cxxLanguageStandard: .cxx17
```
Comment (pepone) [inline Package.swift:212]: Can we use .cxx20 like in C++ macOS builds?

**Review by pepone — APPROVED** [review]

File: Package.swift (hunk 6 above)
Comment (externl) [reply Package.swift:212]: I think so!

**Pull request merged**